File: typescript/hashrule.ts
import * as $$ from "./shell.js";
import * as CACHE from "./data/cache.js";
import Use from "./utils/main.js";

const hashPattern = /#\{[a-z0-9-]+\}/i;

function IMPORT(rule: string, watchUndef = true, source = CACHE.PATH.json.hashrule.path) {
	const primitive = rule;
	const recursionSequence: string[] = [];
	const preview: Record<string, string> = {};

	const response = (
		result: string,
		cause = '',
		message = '',
	) => {
		const E = $$.HashruleError(
			primitive,
			cause,
			source,
			message,
			preview
		);

		return {
			status: message.length === 0,
			result,
			error: E.error,
			diagnostic: E.diagnostic
		};
	};

	let rgxMatch;
	while ((rgxMatch = hashPattern.exec(rule))) {
		const match = rgxMatch[0];
		const key = match.slice(2, -1);
		const replacement = watchUndef
			? CACHE.CLASS.Hashrule[key]
			: (CACHE.CLASS.Hashrule[key] ?? match);
		preview["FROM " + match] = `GETS ${replacement} FROM ${rule}`;

		if (replacement === undefined) {
			return response('', match, "Undefined Hashrule.");
		}
		if (recursionSequence.includes(match)) {
			return response('', match, "Hashrule recursion loop.");
		}
		rule = rule.replace(hashPattern, replacement);

		recursionSequence.push(match);
	}

	return response(rule);
}

function UPLOAD() {
	const errors: string[] = [];
	CACHE.CLASS.Hashrule = CACHE.STATIC.Hashrule;
	const hashrule = { ...CACHE.STATIC.Hashrule };

	Object.keys(hashrule).map((key) => {
		const hash = `#{${key}}`;
		const response = IMPORT(hash);
		if (response.status) {
			hashrule[key] = response.result;
		} else {
			delete hashrule[key];
			errors.push(response.error);
		}
	});

	CACHE.CLASS.Hashrule = hashrule;
	CACHE.DELTA.Manifest.hashrules = hashrule;
	CACHE.DELTA.Report.hashrule = $$.HashruleReport(hashrule, errors);
}

function RENDER(string: string, sourcePath: string) {
	const extended = IMPORT(string, true, sourcePath);
	const snippets = Use.string.zeroBreaks(extended.result, ["&"]);
	const wrappers: string[] = [];

	snippets.forEach(snippet => {
		snippet = snippet.trim();
		const length = snippet.length;
		let wrapper = '', deviance = 0, splAtrule = false;

		for (let i = 0; i < length; i++) {
			const ch = snippet[i];
			if (")}".includes(ch)) { deviance--; }

			if (deviance) {
				wrapper += ch;
			} else {
				switch (ch) {
					case "{":
					case "}":
						wrapper += "";
						break;
					case "@":
						if (wrapper.length) {
							wrapper += " ";
							splAtrule = true;
						}
						wrapper = "@" + wrapper;
						break;
					default:
						wrapper += ch;
				}
			}
			if ("({".includes(ch)) { deviance++; }
		}
		if (wrapper.length) {
			wrappers.push((splAtrule
				? wrapper
					.replace(/width\s*>=/g, "min-width:")
					.replace(/width\s*<=/g, "max-width:")
					.replace(/height\s*>=/g, "min-height:")
					.replace(/height\s*<=/g, "max-height:")
				: wrapper).replace(/\s+/g, " "));
		}
	});

	return {
		wrappers,
		status: extended.status,
		error: extended.error,
		diagnostic: extended.diagnostic
	};
}

function WRAPPER(parentObject: Record<string, object>, keys: string[], childObject: object, parentAtrule = true) {
	const activeKey = keys.shift();
	if (activeKey) {
		const modkey = (parentAtrule || activeKey.startsWith("@")) ? activeKey : ('&' + activeKey);
		if (keys.length) {
			if (!parentObject[modkey]) {
				parentObject[modkey] = {};
				WRAPPER(parentObject[modkey] as Record<string, object>, keys, childObject, activeKey.startsWith("@"));
			}
		} else {
			parentObject[modkey] = childObject;
		}
	}
}; 

export default {
	IMPORT,
	UPLOAD,
	RENDER,
	WRAPPER
};

File: gosource/style/hashrule.go
package style

import (
	X "main/xhell"
	_cache_ "main/cache"
	_types_ "main/types"
	_utils_ "main/utils"
	_regexp_ "regexp"
	_slices_ "slices"
	_strings_ "strings"
)

func Hashrule_Upload() {
	_cache_.Class.Hashrule = _cache_.Static.Hashrule
	hashrule := _cache_.Static.Hashrule
	errors := make([]string, len(hashrule))

	for key := range hashrule {
		var hash = "#{" + key + "}"
		response := Hashrule_Import(hash, _cache_.Path["json"]["hashrule"].Path)
		if response.Status {
			hashrule[key] = response.Result
		} else {
			hashrule[key] = ""
			errors = append(errors, response.Errorstring)
		}
	}

	_cache_.Class.Hashrule = hashrule
	_cache_.Manifest.Hashrule = hashrule
	_cache_.Delta.Report.Hashrule = X.Hashrule_Report(hashrule, errors)
}

type tHashrule_Import_return struct {
	Status      bool
	Result      string
	Errorstring string
	Diagnostic  _types_.Support_Diagnostic
}

func Hashrule_Import(rule string, source string) tHashrule_Import_return {
	primitive := rule
	recursionSequence := make([]string, 0, len(_cache_.Class.Hashrule))
	preview := make(map[string]string, len(_cache_.Class.Hashrule))

	var response = func(
		result string,
		cause string,
		message string,
	) tHashrule_Import_return {
		errorstring, diagnostic := X.Hashrule_Error(
			primitive,
			cause,
			source,
			message,
			preview,
		)

		status := len(message) == 0
		return tHashrule_Import_return{
			Status:      status,
			Result:      result,
			Errorstring: errorstring,
			Diagnostic:  diagnostic,
		}
	}

	var hashpattern = _regexp_.MustCompile(`(?i)#\{[a-z0-9-]+\}`)
	for {
		loc := hashpattern.FindStringIndex(rule)
		if loc == nil {
			break
		}
		match := rule[loc[0]:loc[1]]
		key := match[2 : len(match)-1]
		replacement, found := _cache_.Class.Hashrule[key]
		if !found {
			replacement = match
		}
		preview["FROM "+match] = "GETS " + replacement + " FROM " + rule

		if !found {
			return response("", match, "Undefined Hashrule.")
		}
		if _slices_.Contains(recursionSequence, match) {
			return response("", match, "Hashrule recursion loop.")
		}

		rule = hashpattern.ReplaceAllString(rule, replacement)
		recursionSequence = append(recursionSequence, match)
	}

	return response(rule, "", "")
}

type tHashrule_Render_return struct {
	Wrappers    []string
	Status      bool
	Errorstring string
	Diagnostic  _types_.Support_Diagnostic
}

func Hashrule_Render(
	str string,
	src string,
) tHashrule_Render_return {
	const wrapperlimit = 12
	extended := Hashrule_Import(str, src)
	snippets := _utils_.String_ZeroBreaks(extended.Result, []rune{'&'})[:wrapperlimit]
	wrappers := make([]string, 0, wrapperlimit)

	for _, snippet := range snippets {
		var wrapper _strings_.Builder

		snippet = _strings_.TrimSpace(snippet)
		length := len(snippet)
		deviance := 0
		splAtrule := false
		holdAtRune := false

		for i := 0; i < length; i++ {
			ch := rune(snippet[i])
			if ch == ')' || ch == '}' || ch == ']' {
				deviance--
			}

			if deviance > 0 {
				wrapper.WriteRune(ch)
			} else {
				switch ch {
				case '{':
				case '}':
				case '@':
					if wrapper.Len() > 0 {
						wrapper.WriteRune(' ')
						splAtrule = true
					}
					holdAtRune = true
				default:
					wrapper.WriteRune(ch)
				}
			}
			if ch == '(' || ch == '{' || ch == '[' {
				deviance++
			}
		}

		if wrapper.Len() > 0 {
			wrapped := wrapper.String()
			if holdAtRune {
				wrapped = "@" + wrapped
			}
			if splAtrule {
				wrapped = _regexp_.MustCompile(`width\s*>=`).ReplaceAllString(wrapped, "min-width:")
				wrapped = _regexp_.MustCompile(`width\s*<=`).ReplaceAllString(wrapped, "max-width:")
				wrapped = _regexp_.MustCompile(`height\s*>=`).ReplaceAllString(wrapped, "min-height:")
				wrapped = _regexp_.MustCompile(`height\s*<=`).ReplaceAllString(wrapped, "max-height:")
			}
			wrapped = _utils_.String_Minify(wrapped)
			wrappers = append(wrappers, wrapped)
		}
	}

	return tHashrule_Render_return{
		Wrappers:    wrappers,
		Status:      extended.Status,
		Errorstring: extended.Errorstring,
		Diagnostic:  extended.Diagnostic,
	}
}

func Hashrule_Wrapper(parentObject map[string]any, keys []string, childObject any) {
	if len(keys) == 0 {
		return
	}

	activeKey := keys[0]
	keys = keys[1:]

	if len(keys) > 0 {
		sub, ok := parentObject[activeKey]
		if !ok {
			sub = make(map[string]any)
			parentObject[activeKey] = sub
		}

		if m, ok := sub.(map[string]any); ok {
			Hashrule_Wrapper(m, keys, childObject)
		}
	} else {
		parentObject[activeKey] = childObject
	}
}
